dimod/include/dimod/utils.h

```diff
@@ -27,7 +29,85 @@ namespace utils {
         return ub.first < v;
     }
 
+    // zip_sort is a modification of the code found here :
+    // https://www.geeksforgeeks.org/iterative-quick-sort/
+
+    /// Sort two vectors, using `control` to provide the ordering.
+    template <class T1, class T2>
+    void zip_sort(std::vector<T1>& control, std::vector<T2>& response) {
+        assert(control.size() == response.size());
+        std::size_t length = control.size();
+
+        if (length < 2) {
+            return;
+        } else if (length > LLONG_MAX) {
+            throw std::logic_error(
+                    "Length of arrays are too big for sorting. Numerical "
+                    "overflow will occur.");
+        }
+
+        std::int64_t* stack =
+                (std::int64_t*)malloc(length * sizeof(std::int64_t));
+        std::int64_t top = -1;
+        std::int64_t low = 0;
+        std::int64_t high = length - 1;
+
+        stack[++top] = low;
+        stack[++top] = high;
+
+        while (top >= 0) {
+            high = stack[top--];
+            low = stack[top--];
+
+            std::int64_t pivot_choice_index = (low + high + 1) / 2;
+            if (pivot_choice_index != high) {
+                T1 control_temp = control[pivot_choice_index];
```

Comment:
I wonder how in the world this code survived all the randomized tests. , I am swapping the control variables, but not the response variables. here.  just need to add the following lines, but I will follow up with more testing after making the changes myself 

```
            T2 response_temp = response[pivot_choice_index];
            response[pivot_choice_index] = response[high];
            response[high] = response_temp;
```

Reply:
Found the reason why my tests never caught it. I think it will seem familiar to you 

```
112         for (int i = 0; i < num_elements; i++) {
113             if (v_pair[i].first != control[i] &&
114                 v_pair[i].second != response[i]) {
115                 std::cout << "Problem Pairs not sorted properly" << std::endl;
116             }
117         }
```
 
The && should be a ||. :)

Reply:
The suggested code snipped fixed it, I also pushed my fix in the repo I originally had.